crates/matrix-sdk-ui/src/notification_client.rs

```diff
@@ -84,94 +94,317 @@ impl NotificationClient {
-            is_direct_message_room: room.is_direct().await?,
-            is_room_encrypted: room.is_encrypted().await.ok(),
-            joined_members_count: room.joined_members_count(),
-            is_noisy,
+        let push_actions = match &raw_event {
+            RawNotificationEvent::Timeline(timeline_event) => {
+                // Timeline events may be encrypted, so make sure they get decrypted first.
+                if let Some(timeline_event) =
+                    self.maybe_retry_decryption(&room, &timeline_event).await?
+                {
+                    raw_event = RawNotificationEvent::Timeline(timeline_event.event.cast());
+                    timeline_event.push_actions
+                } else {
+                    room.event_push_actions(&timeline_event).await?
+                }
+            }
+            RawNotificationEvent::Invite(invite_event) => {
+                // Invite events can't be encrypted, so they should be in clear text.
+                room.event_push_actions(&invite_event).await?
+            }
         };
 
-        Ok(Some(item))
+        if let Some(push_actions) = &push_actions {
+            if self.filter_by_push_rules && !push_actions.iter().any(|a| a.should_notify()) {
+                return Ok(None);
+            }
+        }
+
+        Ok(Some(NotificationItem::new(false, &room, &raw_event, push_actions.as_deref()).await?))
+    }
+
+    async fn handle_timeline_event(
+        &self,
+        room: &Room,
+        mut timeline_event: TimelineEvent,
+        sync_members: bool,
+    ) -> Result<Option<NotificationItem>, Error> {
+        if let Some(decrypted_event) =
+            self.maybe_retry_decryption(room, timeline_event.event.cast_ref()).await?
+        {
+            timeline_event = decrypted_event;
+        }
+
+        if self.filter_by_push_rules
+            && !timeline_event
+                .push_actions
+                .as_ref()
+                .is_some_and(|actions| actions.iter().any(|a| a.should_notify()))
+        {
+            return Ok(None);
+        }
+
+        Ok(Some(
+            NotificationItem::new(
+                sync_members,
+                room,
+                &RawNotificationEvent::Timeline(timeline_event.event.cast()),
+                timeline_event.push_actions.as_deref(),
+            )
+            .await?,
+        ))
+    }
+
+    /// Retrieve a notification using a /context query.
+    ///
+    /// This is for clients that are already running other sliding syncs in the
+    /// same process, so that most of the contextual information for the
+    /// notification should already be there.
+    pub async fn get_notification_with_context(
+        &self,
+        room_id: &RoomId,
+        event_id: &EventId,
+    ) -> Result<Option<NotificationItem>, Error> {
+        tracing::info!("fetching notification event with a /context query");
+
+        // XXX Will this work for invites? A sliding sync may be running in the
```

Comment:
Yeah, that's racy. Like discussed in the Matrix room, this should be called once a `SS` response has put the room into the memory store in the case the `SS` response didn't contain the event.

But let's do that in a separate PR and once the clients confirm that this is all fine and well.